### PR TITLE
Support for running in a non-Node.js environment

### PR DESCRIPTION
**Pull request by oyisre**
Running outside of Node.js requires Web Workers to be defined at `window.Worker`, and it currently expects to find a worker script at "./cql.worker.js". Maybe it would be better to pass the worker script URL/path as an optional parameter to applyPlan?

This has only been tested using Webpack 5 btw, but let me know if you'd like to see that config file as well

`src/main.js:9` is messy, but I had no particular reason against using an "if" there, just preference. We just need to avoid running those functions when not in Node

I also threw in a couple bug fixes for some problems we ran into

### CODE REVIEW

File: src/simpleResolver.js [hunk 1]
```diff
@@ -115,6 +115,12 @@ function compareSemver(a,b) {
   // If either a or b do not have a version element, set a default.
   const aVersion = a.version ?? '0.0.0';
   const bVersion = b.version ?? '0.0.0';
+  if (aVersion.split('.').length < 3) {
+    aVersion = coerce(aVersion)?.raw || '0.0.0';
```
Comment (dwwinters) [inline src/simpleResolver.js:123]: Where does `coerce()` get defined?

File: src/simpleResolver.js [hunk 2]
```diff
@@ -89,7 +89,8 @@ export function simpleResolver(input = null) {
             && (rsrc.id == resourceId || rsrc.url == url);
         }).sort(compareSemver); 
         // NOTE: We're grabbing the first index which should be latest version.
-        resolvedReference = [sortedFhirJson.shift()];
+        const shifted = sortedFhirJson.shift();
+        resolvedReference = shifted !== undefined ? [shifted] : [];
```
Comment (dwwinters) [inline src/simpleResolver.js:106]: Great catch.  I'm adding a test to cover this situation.

Comment (dwwinters) [inline src/simpleResolver.js:106]: Great catch.  I'm adding a test to cover this situation.

File: src/simpleResolver.js [hunk 3]
```diff
@@ -42,6 +42,9 @@ export function simpleResolver(input = null) {
     // Is this a bundle?
     if (fhirJson?.resourceType == 'Bundle') {
       // Flatten out the entries
+      fhirJson = fhirJson.entry !== undefined ?
+        fhirJson.entry.map(r => r.resource) :
+        [];
       fhirJson = fhirJson?.entry.map(r => r.resource);
```
Comment (dwwinters) [inline src/simpleResolver.js:49]: This line is redundant now.  I'll remove it.

Comment (dwwinters) [inline src/simpleResolver.js:49]: This line is redundant now.  I'll remove it.

File: src/simpleResolver.js [hunk 4]
```diff
@@ -42,6 +42,9 @@ export function simpleResolver(input = null) {
     // Is this a bundle?
     if (fhirJson?.resourceType == 'Bundle') {
       // Flatten out the entries
+      fhirJson = fhirJson.entry !== undefined ?
```
Comment (dwwinters) [inline src/simpleResolver.js:46]: Added a test for this.

Comment (dwwinters) [inline src/simpleResolver.js:46]: Added a test for this.

**Comment by dwwinters** [comment]
Thanks for the PR @oyisre.  Looks good; I agree the handling of the node vs browser environment is a little kludgey.  But I don't see a better way other than to pull the web work initialization out and ask the user to set it up themselves and then pass it in (via `aux`?).

I did have some questions that I included via comments on certain parts of the code; see above.  My one open question I think is about the use of the `coerce()` function in the version comparison code.  I think this isn't necessary so I commented it out.

I've created a copy of your branch [here](https://github.com/ccsm-cds-tools/encender/commits/PuraJuniper-browser-support).  On this branch you'll see I've removed code that is now redundant and added more tests to validate the bug fixes in your PR.  I also brought over additional changes you made on [another branch from your fork](https://github.com/PuraJuniper/encender/pull/3), which includes support for async resolver() functions and a fix for the generated RequestGroup.action[].resource element.  Updated the tests to account for those.

Take a look at the [modifications](https://github.com/ccsm-cds-tools/encender/commits/PuraJuniper-browser-support) I made and let me know what you think.  If we're happy with it, I can either merge that new branch into main or you can merge it into your branch and I can merge the PR.  Let me know either way.

**Comment by dwwinters** [comment]
Oh, and we should update the README with instructions for webpack too.  But I can do that after we merge this.

**Comment by dwwinters** [comment]
@oyisre @goelalex I've created a [new branch](https://github.com/ccsm-cds-tools/encender/tree/alt-PuraJuniper-browser-support) that takes a different approach to what you've proposed for supporting browser applications.

Instead of trying to automatically detect Node vs Browser and construct the worker accordingly, I think it may make more sense to ask users to pass in a factory function that returns constructed workers that are appropriate for their environment.  Example using web workers:

```js
const WorkerFactory = () => {
  return new Worker(new URL('../node_modules/cql-worker/src/cql.worker.js', import.meta.url))
};
```

`WorkerFactory` then gets passed into Encender via the `aux` input parameter.  If no factory is passed in, then the default would be to assume a Node environment.

This approach has the advantage that it is more straightforward when it comes to configuring your bundling tool (e.g., webpack).  I ran into some issues with this PR when using a simple CRA with ejected configuration; the webpack configuration needed a lot more fiddling to get encender to work.  With [this alternative approach](https://github.com/ccsm-cds-tools/encender/tree/alt-PuraJuniper-browser-support), no additional configuration was needed aside from polyfilling the node libraries.

File: src/simpleResolver.js (hunk 1 above)
Comment (oyisre) [reply src/simpleResolver.js:123]: It should be imported from the "semver" library, my bad!

As a side note for why I added it in the first place: the FHIR spec doesn't enforce semantic versioning, and I've encountered runtime errors due to resources with non-compliant versions like "1.0"

**Comment by oyisre** [comment]
The new branch looks good to me! I agree, it makes sense to pass in environment differences through a parameter. I'll be able to test it out in our app this week and provide feedback

Thanks for adding the tests 🙂

**Comment by oyisre** [comment]
I've tested [the new branch](https://github.com/ccsm-cds-tools/encender/tree/alt-PuraJuniper-browser-support) with our in-browser app and it works great! No changes are needed from our end, so I support merging that branch in if that works with you

**Comment by oyisre** [comment]
I've tested [the new branch](https://github.com/ccsm-cds-tools/encender/tree/alt-PuraJuniper-browser-support) with our in-browser app and it works great! No changes are needed from our end, so I support merging the branch in if that works with you

**Comment by dwwinters** [comment]
@oyisre I'll close this PR and merge the other branch.  Thanks for the contribution!!! :)

**Pull request merged**